### PR TITLE
IBM Cloud loadbalancer implementation

### PR DESCRIPTION
**Pull request by ericvn**
Signed-off-by: Eric Van Norman <ericvn@us.ibm.com>

### CODE REVIEW

File: docs/controller/ingress/example.yml [hunk 1]
```diff
@@ -16,9 +16,9 @@
 # Verify that the route has been added in the simulator/lb1:
 # infrakit simulator/lb1 routes ls
 #
-# FRONTEND PORT    PROTOCOL   BACKEND PORT  CERT
-# 7777             TCP        7777  <-- from the swarm service
-# 80               http       8080  <-- from the static route in the config
+# FRONTEND PORT    FRONTEND PROTOCOL      BACKEND PORT      BACKEND PROTOCOL        CERT
```
Comment (kaufers) [inline docs/controller/ingress/example.yml]: Why are these changes here? Is this file out of date? Can you put something in the commit message explaining why this is updated? It seems like that the `Protocol` case is the main delta here (`http` -> `HTTP`).

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+
+	// Iterate over the list of members to add checking if they already exist in the loadbalancer.
+	// If not, add member.
+	for _, id := range ids {
+		ip := string(id)
+		ipFound := false
+		for _, member := range lb.Members {
+			if *member.Address == ip {
+				ipFound = true
+				lbLogger.Error("RegisterBackends", "name", l.name, "V", debugV, "backend already exists", ip)
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Why is this an error? Won't this be hit if there is a rolling update and now we have a new member?

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+	// Get current list of backend members
+	lb, err := l.lbService.Mask("members").GetLoadBalancer(&l.uuid)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over the list of members to add checking if they already exist in the loadbalancer.
+	// If not, add member.
+	for _, id := range ids {
+		ip := string(id)
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: So the `ids` that are passed in are actually IPv4 addresses? If so, can you add a comment that explains this?

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+		attempt := 0
+		for {
+			attempt++
+			_, err = l.lbMemberService.AddLoadBalancerMembers(&l.uuid, membersToAdd)
+			if err == nil {
+				break
+			}
+			// Check for the lb is in state UPDATE_PENDING. (HTTP 500)
+			if strings.Contains(err.Error(), "UPDATE_PENDING") && attempt <= updateRetryCount {
+				time.Sleep(updateRetryTime)
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Can we add in some logging here that exposes this error case? I like that we are handling it but it would be nice to get visibility into how often this happens.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+		ip := string(id)
+		for _, member := range lb.Members {
+			if *member.Address == ip {
+				membersToRemove = append(membersToRemove, *member.Uuid)
+				ipFound = true
+				break
+			}
+		}
+		if ipFound == false {
+			lbLogger.Error("DeregisterBackends", "name", l.name, "V", debugV, "backend not found", ip)
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Again why is this an error? Shouldn't be the normal case when the group member changes (like on a rolling update)?

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+		attempt := 0
+		for {
+			attempt++
+			_, err = l.lbMemberService.DeleteLoadBalancerMembers(&l.uuid, membersToRemove)
+			if err == nil {
+				break
+			}
+			// Check for the lb is in state UPDATE_PENDING. (HTTP 500)
+			if strings.Contains(err.Error(), "UPDATE_PENDING") && attempt <= updateRetryCount {
+				time.Sleep(updateRetryTime)
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Same comment on the logging.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+				return nil, err
+			}
+		}
+	}
+
+	lbLogger.Debug("DeregisterBackends", "name", l.name, "V", debugV, "backends", membersToRemove)
+	return lbResult("deregister"), nil
+}
+
+// Backends returns a list of backends
```
Comment (kaufers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Can we explain that we are using the IP address as a common correlation key? Even though this func is returning an `[]instance.ID`, the value is simply the IP address of each member, right?

File: pkg/run/v0/ibmcloud/ibmcloud.go [hunk 8]
```diff
@@ -39,11 +49,38 @@ type VolumeAuth struct {
 	VolumeID int
 }
 
+// LB is custom type for IBM Cloud loadbalancer definition
+type LB struct {
+	// LBName is the name of the IBM Cloud loadbalancer
+	LBName string
+
+	// LBUUID is the UUID of the IBM cloud loadbalancer
```
Comment (kaufers) [inline pkg/run/v0/ibmcloud/ibmcloud.go]: s/IBM cloud/IBM Cloud

File: pkg/run/v0/ibmcloud/ibmcloud.go [hunk 9]
```diff
@@ -76,11 +114,29 @@ func Run(scope scope.Scope, name plugin.Name,
 		authInstPlugin = ibmcloud_auth_inst.NewVolumeAuthPlugin(options.Username, options.APIKey, options.VolumeAuth.VolumeID)
 	}
 
+	l4Map := map[string]loadbalancer.L4{}
+	for _, lb := range options.LBs {
+		var lbPlugin loadbalancer.L4
+		lbPlugin, err = ibmcloud_loadbalancer.NewIBMCloudLBPlugin(options.Username, options.APIKey, lb.LBName, lb.LBUUID)
+		if err != nil {
+			return
```
Comment (kaufers) [inline pkg/run/v0/ibmcloud/ibmcloud.go]: Should we log an error here?

File: docs/controller/ingress/example.yml (hunk 1 above)
Comment (ericvn) [reply docs/controller/ingress/example.yml]: They've simply been sitting in my branch waiting to be put in as David said to put them in with my next PR. I could certainly push them in another PR if that is warranted.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 2 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: I though I put this error message into the code pulling from either the simulator or AWS load balancer code, but can't find that now. This code logs an error message that the backend was already registered, but doesn't change any behavior. I'm fine with removing this check entirely, or downgrading the message to a warning/info/debug.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 3 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Added into the function remarks.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 4 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Done in the several places we have this code.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 5 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Similar to the above comment.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 6 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Done.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 7 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Correct.

Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: I'll add a bigger remark that we are using the private IP address of the worker instead of the worker name. This was because there wasn't a field to map worker name to ip within the LBaaS APIs.

File: pkg/run/v0/ibmcloud/ibmcloud.go (hunk 9 above)
Comment (ericvn) [reply pkg/run/v0/ibmcloud/ibmcloud.go]: I find the logging inconsistent across plugins, so adding it here.

File: pkg/run/v0/ibmcloud/ibmcloud.go (hunk 8 above)
Comment (ericvn) [reply pkg/run/v0/ibmcloud/ibmcloud.go]: Done

File: docs/controller/ingress/example.yml [hunk 10]
```diff
@@ -58,9 +58,9 @@ properties:
     # Here we have a static route that is always present.
     Routes:
       - LoadBalancerPort: 80
-        LoadBalancerProtocol: https
+        LoadBalancerProtocol: HTTP
```
Comment (chungers) [inline docs/controller/ingress/example.yml]: Did this become case sensitive at some point?   Curious why you'd change it to all caps.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+		}
+		if len(cert) == 1 {
+			certID = *cert[0].Id
+		} else {
+			return nil, fmt.Errorf("Cannot identify an unique certificate with common name '%s'", *route.Certificate)
+		}
+	}
+
+	// InfraKit doesn't define a method, so default to round robin
+	lbMethod := "ROUNDROBIN"
```
Comment (chungers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Is this something that we can add to the `Route` struct?  Different L4 implementations will have different strings but if this is just some opaque value that's passed on to the L4 then we can avoid this hardcoding here.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+	lbLogger.Debug("NewIBMCloudLBPlugin", "V", debugV, "Name", lbName)
+	lb := &ibmcloudlb{
+		softlayerUsername: username,
+		softlayerAPIKey:   apikey,
+		name:              lbName,
+		uuid:              lbUUID,
+	}
+
+	// Hardcoded IBM Cloud endpoints
+	lb.iamEndpoint = "https://iam.ng.bluemix.net"
```
Comment (chungers) [inline pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: It might be useful to define a few exported / public constants and not have these values buried in here.

File: docs/controller/ingress/example.yml (hunk 10 above)
Comment (ericvn) [reply docs/controller/ingress/example.yml]: Let me go back and look at this one as I made it a while ago. I recall at the time that the check between what is in the yml and what the Routes API returned for protocol was case sensitive, and this might have been fixed. I moved to the Protocol case in protocol.go.

File: docs/controller/ingress/example.yml (hunk 1 above)
Comment (chungers) [reply docs/controller/ingress/example.yml]: Let's remove changes to this file in this PR...

Comment (ericvn) [reply docs/controller/ingress/example.yml]: Removed.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 11 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: The specification of the method could be added. For information, the IBM LBaaS LB supports 3 methods: Round Robin, Weighted Round Robin (would also require the addition of a weight on each backend), and Least Connection.

File: pkg/provider/ibmcloud/plugin/loadbalancer/lb.go (hunk 12 above)
Comment (ericvn) [reply pkg/provider/ibmcloud/plugin/loadbalancer/lb.go]: Done.
